src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -11,207 +12,171 @@ namespace System.Buffers
+                {
+                    next = GetBufferCrossSegment(ref startMemoryList, end, ref length);
+                }
+            }
+            else if (type != SequenceType.Empty && startSegment != endSegment)
+            {
+                ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+            }
 
             switch (type)
             {
                 case SequenceType.MemoryList:
-                    var segment = (IMemoryList<T>)start.Segment;
-                    Memory<T> bufferSegmentMemory = segment.Memory;
-                    int currentEndIndex = bufferSegmentMemory.Length;
-
-                    if (segment == end.Segment)
-                    {
-                        currentEndIndex = endIndex;
-                        next = default;
-                    }
-                    else
-                    {
-                        IMemoryList<T> nextSegment = segment.Next;
-                        if (nextSegment == null)
-                        {
-                            if (end.Segment != null)
-                            {
-                                ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
-                            }
-
-                            next = default;
-                        }
-                        else
-                        {
-                            next = new SequencePosition(nextSegment, 0);
-                        }
-                    }
-
-                    data = bufferSegmentMemory.Slice(startIndex, currentEndIndex - startIndex);
-                    return true;
-
-                case SequenceType.OwnedMemory:
-                    var ownedMemory = (OwnedMemory<T>)start.Segment;
-                    if (ownedMemory != end.Segment)
-                    {
-                        ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
-                    }
-
-                    data = ownedMemory.Memory.Slice(startIndex, endIndex - startIndex);
-                    next = default;
-                    return true;
-
+                    data = startMemoryList.Memory.Slice(startIndex, length);
+                    break;
                 case SequenceType.Array:
-                    var array = (T[])start.Segment;
-
-                    if (array != end.Segment)
-                    {
-                        ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
-                    }
-
-                    data = new Memory<T>(array, startIndex, endIndex - startIndex);
-                    next = default;
-                    return true;
-
+                    data = new ReadOnlyMemory<T>((T[])startSegment, startIndex, length);
+                    break;
+                case SequenceType.OwnedMemory:
+                    data = ((OwnedMemory<T>)startSegment).Memory.Slice(startIndex, length);
+                    break;
+                case SequenceType.String:
```

Comment:
`if (typeof(T) == typeof(string))` ?

Reply:
`typeof(T) == typeof(char)`

Reply:
Right, I meant,  should we put this `case` behind the `if`? Would improve method size for `byte` case.

Reply:
ah good point